### PR TITLE
Typescript Plumbing

### PR DESCRIPTION
**Pull request by jens-ox**
This prepares the build chain for TypeScript. Now, files can "simply" be renamed to `.ts` and everything should work, which I will do in a follow-up PR.

Note that I had to do some changes to the testing setup and the code, notably:

- Added some async/await due to race conditions causing flaky tests. CI is successful, but I assume that there's still a bit of flakiness left. That should be fixed by making sure that async/await is handled correctly across the codebase, which will be done in the follow-up TypeScript PR.
- Switched the Puppeteer headless mode to `new` to get rid of discrepancies in the testing setup between Mac and Win/Ubuntu.
- Removed the UMD bundle test, as there's no UMD bundle.

[Successful CI run in the fork](https://github.com/jens-ox/niivue/actions/runs/6990750353?pr=6)

### CODE REVIEW

**Review by hanayik — COMMENTED** [review]
@jens-ox, NiiVue has a feature that allows it to save a scene. This is useful for configuring a canvas and then downloading a self contained html file usable offline, with no dependencies. This uses some features to inject the NiiVue bundle into the generated html file. You can see a demo [here](https://niivue.github.io/niivue/features/save.html.html). This PR (which is full of great improvements )has removed the vite config that allowed the two stage build to inject (in compressed form) NiiVue into itself.

**Comment by jens-ox** [comment]
@hanayik ah, that makes sense!

To be honest, I'm not sure if that should be a feature of the core library.

- It greatly increases the bundle size of NiiVue.
- It will only work in scenarios where NiiVue is used as a sole dependency.
- The same result can be achieved by e.g. running [vite-plugin-singlefile](https://www.npmjs.com/package/vite-plugin-singlefile) on the application server.
- It makes the inclusion of all of NiiVue's dependencies mandatory, which means that a proper ES Module build with tree-shakable dependencies isn't really possible.

While the feature itself is certainly very useful, it might make more sense to have it in e.g. https://github.com/niivue/niivue-apps.

**Comment by hanayik** [comment]
@jens-ox, all good points. I think this is something that would be another good topic to bring up in the NiiVue group meeting on the 29th. If we do remove the feature, we should at least document how the same behaviour can be achieved (by some application in niivue-apps).

**Comment by hanayik** [comment]
@jens-ox, it would be good to rebase on main to pull in the curly bracket eslint config and updates.

**Comment by jens-ox** [comment]
@hanayik done 👍🏼

**Comment by jens-ox** [comment]
While it's not ideal, I've increased the Jest timeouts for now, which allows all tests to run through (see https://github.com/jens-ox/niivue/actions/runs/7025741507). I don't really know why `test.shader.sobel.js` loves to fail on Mac (which I can't reproduce on my machine locally) - I assume it's some weird race condition. I hope that the extended timeout won't be necessary after a migration to TypeScript.

**Comment by hanayik** [comment]
@jens-ox, I'm not sure, but it might be the case that parallel test running is causing some of the flakiness you observed in CI but not locally. Perhaps the machines are under-resourced? In general Puppeteer has been flaky with our tests, and was one of the motivators to switch to Playwright.

**Comment by jens-ox** [comment]
@hanayik makes sense. As the Playwright PR is already open, I wouldn't investigate any further here.

**Comment by hanayik** [comment]
@jens-ox , is this still a WIP, or ready for review and merge?

**Comment by jens-ox** [comment]
@hanayik from my end this is ready for review, but it removes the feature to save as static HTML. If that is ok, then it can be merged.

**Comment by jens-ox** [comment]
@hanayik I forgot to update the demo imports, so I fixed that - sorry! Should've been the last commit on this PR, hopefully ☺️

**Comment by hanayik** [comment]
@jens-ox and @neurolabusc, how about this proposal for the "saveHTML" feature:

- we change [`saveHTML`](https://github.com/niivue/niivue/blob/f4eb05ac039994697f509d2a89ea11de04c05bae/src/niivue/index.js#L3357) to: `saveHTML(fileName, htmlString)`
- `htmlString` will need to be created by the consuming application or web page
- `generateLoadDocumentJavaScript(canvasId)` will change to `generateLoadDocumentJavaScript(canvasId, niivueModuleUrl)` where `niivueModuleUrl` is the path to a prebundled niivue. The file at `niivueModuleUrl` will be read as text from the url, then saved as a string. 
- the downstream developer will also need to produce the niivue bundle. For us, we can do this as part of a separate build unrelated to publishing so that our demos are unchanged in functionality. 

To keep our live demos unchanged, we would change from this in `save.html.html`

```
    import * as niivue from "./niivue.es.js";
    function saveAsHtml() {
      nv1.saveHTML("page.html");
    }
    var volumeList1 = [
      {
        url: "../images/mni152.nii.gz",
        colormap: "gray",
        visible: true,
        opacity: 1
      },
      {
        url: "../images/hippo.nii.gz",
        colormap: "red",
        visible: true,
        opacity: 1
      }
    ];

    // assign our event handler
    var button = document.getElementById("save");
    button.onclick = saveAsHtml;
    var nv1 = new niivue.Niivue();
    nv1.attachTo("gl1");
    nv1.loadVolumes(volumeList1);
    nv1.setSliceType(nv1.sliceTypeRender);
```

to:

```
    import * as niivue from "./niivue.es.js";
    function saveAsHtml() {
      const javaScript = nv1.generateLoadDocumentJavaScript(canvasId, './niivue.es.js')
      const htmlString = `inject ${javaScript} here. some useful html here., blah blah blah.`
      nv1.saveHTML("page.html", htmlString);
    }
    var volumeList1 = [
      {
        url: "../images/mni152.nii.gz",
        colormap: "gray",
        visible: true,
        opacity: 1
      },
      {
        url: "../images/hippo.nii.gz",
        colormap: "red",
        visible: true,
        opacity: 1
      }
    ];

    // assign our event handler
    var button = document.getElementById("save");
    button.onclick = saveAsHtml;
    var nv1 = new niivue.Niivue();
    nv1.attachTo("gl1");
    nv1.loadVolumes(volumeList1);
    nv1.setSliceType(nv1.sliceTypeRender);
```

**Comment by neurolabusc** [comment]
@hanayik this sounds good. Can we also have the build scripts generate a suitable base64 bundles that are available for the latest version `https://niivue.github.io/niivue/features/niivue.b64` as well as all future versions:`https://unpkg.com/@niivue/niivue@0.82.0/dist/niivue.b64`. Furthermore, if the `htmlString` argument is not provided it assume the latest version from the web?

**Comment by jens-ox** [comment]
@hanayik @neurolabusc I think this can be done even simpler - instead of pulling the UMD bundle out of NiiVue itself, we expose a `niivue` string parameter on `generateLoadDocumentJavaScript` (and transitively `saveHTML`). Users can then pass a NiiVue bundle (or a NiiVue bundle with extra dependencies).

For the b64 target: I agree that it makes sense to publish a compressed version for this use-case alongside the ES Module build. I would not recommend to publish it as a Base64-encoded string though, as it obviously won't affect the main NiiVue bundle size and will be decoded anyway before being placed in the HTML output.

tl;dr:
- generate minified module with inlined dependencies alongside main ES Module build
- expose `niivue` parameter on `saveHTML` and `generateLoadDocumentJavaScript` to allow users to inject the NiiVue bundle

**Comment by jens-ox** [comment]
@hanayik @neurolabusc

I toyed around with different approaches and landed on the following:

- export a `dist/index.min.js` alongside the main `dist/index.js`.
- `dist/index.min.js` includes NiiVue as URL encoded string.
- `saveHTML` and related methods have a `esm` property for passing JavaScript.
- Users can either pass the NiiVue instance from `index.min.js` (which they can import from `@niivue/niivue/min`, see the `package.json`) or pass their own bundle.

Imo this decently solves the three problems:
- **Bundle Size**: the compressed instance is not part of the main ES Module bundle, therefore reducing the main bundle size by around 3MB (or a bit less if Base64-encoded).
- **Customization**: users can pass a custom NiiVue instance (or other JS libraries) to the saved HTML.
- **Compile Step Complexity**: while `npm run build` now takes around 2s instead of around 120ms (on my machine:tm:), there is no two-stage self-injection anymore; instead, two separate bundles are created. `npm run build` indirectly runs `bundle.js`, which does nothing besides invoking `tsup` and writing the url-encoded result to another file.

All demos using the export functionality are now working.

---

@neurolabusc loading the current NiiVue version from the web if no `esm` parameter is provided would even further simplify the code, as we could remove the URL-encoding shenanigans and publish a normal bundle to e.g. [Unpkg](https://www.unpkg.com/). The reason for URL-encoding it is that the browser can't import a file as string. Loading it from a CDN would mitigate that.

I would suggest to further simplify this in a future PR once a bundle is pushed to Unpkg.

**Comment by jens-ox** [comment]
@hanayik [the fork pipeline is green](https://github.com/jens-ox/niivue/actions/runs/7045276351), so this this should be once again ready for review 😅 

As this is a major restructuring, the possibility of something breaking without being noticed is quite high. If that should be the case, I will fix those issues as soon as possible.

File: src/index.html [hunk 1]
```diff
@@ -34,9 +34,7 @@
     </main>
     <footer id="location">&nbsp;</footer>
     <script type="module" async>
-      import { Niivue } from "./niivue";
-      import { NVImage } from "./nvimage";
-      import { NVMesh } from "./nvmesh";
+      import { Niivue, NVImage, NVMesh } from "../dist/index.js";
```
[the remark this answers is not in the document]
Comment (jens-ox) [reply src/index.html]: Fixed 👍🏼

**Comment by neurolabusc** [comment]
@jens-ox can you change [nvimage](https://github.com/jens-ox/niivue/blob/83899e8049a252b9875601d3828fb30c1f19c5f3/src/nvimage/index.js#L2749) from:

```
          if (!isNifti1) {
            dataBuffer = null
          } else {
```
to read
```
          if (!isNifti1) {
            dataBuffer = await this.readFileAsync(file)
          } else {
```

This will resolve [issue 751](https://github.com/niivue/niivue/issues/751), albeit `limitFrames4D` now only applies to NIfTI 4D volumes and not NRRD data. This seems like a reasonable tradeoff for the other gains of this PR.

File: src/nvmesh-loaders.js [hunk 2]
```diff
@@ -786,11 +794,7 @@ export class NVMeshLoaders {
       pos++ // skip null termination
       return new TextDecoder().decode(buffer.slice(startPos, pos - 1))
     } // readStr: read variable length string
-    const scalars = new Float32Array(nvert * nMaps)
-    const maps = []
-    // read Name of SRF
-    let pos = 9
-    // let filenameSRF = readStr();
+
```
Comment (neurolabusc) [inline src/nvmesh-loaders.js:797]: @jens-ox lets add an underscore as you [suggest here](https://github.com/niivue/niivue/issues/768)
```
let _filenameSRF = readStr();
```

**Comment by neurolabusc** [comment]
@hanayik I am done with my review, and happy to have this merged once @jens-ox adds the underscore.

**Review by hanayik — APPROVED** [review]
